common/raw_hashtable.h

```diff
@@ -1399,79 +1483,78 @@ BaseImpl<InputKeyT, InputValueT, InputKeyContextT>::GrowAndInsert(
 template <typename InputBaseT, ssize_t SmallSize>
 TableImpl<InputBaseT, SmallSize>::TableImpl(const TableImpl& arg)
     : BaseT(arg.alloc_size(), arg.growth_budget_, SmallSize) {
+  CARBON_DCHECK(arg.alloc_size() > 0) << "Cannot copy from a moved-from table!";
   CARBON_DCHECK(arg.small_alloc_size_ == SmallSize);
+  CARBON_DCHECK(!arg.is_small() || arg.alloc_size() == SmallSize);
+  CARBON_DCHECK(this->small_alloc_size_ == SmallSize);
 
-  ssize_t local_size = arg.alloc_size();
-
-  if (SmallSize > 0 && arg.is_small()) {
-    CARBON_DCHECK(local_size == SmallSize);
-    this->storage() = small_storage();
-  } else {
-    this->storage() = BaseT::Allocate(local_size);
-  }
-
-  // Preserve which slot every entry is in, including tombstones in the
-  // metadata, in order to copy into the new table's storage without rehashing
-  // all of the keys. This is especially important as we don't have an easy way
-  // to access the key context needed for rehashing here.
-  uint8_t* local_metadata = this->metadata();
-  EntryT* local_entries = this->entries();
-  const uint8_t* local_arg_metadata = arg.metadata();
-  const EntryT* local_arg_entries = arg.entries();
-  memcpy(local_metadata, local_arg_metadata, local_size);
+  SetupStorage();
+  this->CopySlotsFrom(arg);
+}
 
-  for (ssize_t group_index = 0; group_index < local_size;
-       group_index += GroupSize) {
-    auto g = MetadataGroup::Load(local_arg_metadata, group_index);
-    for (ssize_t byte_index : g.MatchPresent()) {
-      local_entries[group_index + byte_index].CopyFrom(
-          local_arg_entries[group_index + byte_index]);
+template <typename InputBaseT, ssize_t SmallSize>
+auto TableImpl<InputBaseT, SmallSize>::operator=(const TableImpl& arg)
+    -> TableImpl& {
+  CARBON_DCHECK(arg.alloc_size() > 0) << "Cannot copy from a moved-from table!";
+  CARBON_DCHECK(arg.small_alloc_size_ == SmallSize);
+  CARBON_DCHECK(!arg.is_small() || arg.alloc_size() == SmallSize);
+  CARBON_DCHECK(this->small_alloc_size_ == SmallSize);
+
+  // We have to end up with an allocation size exactly equivalent to the
+  // incoming argument to avoid re-hashing every entry in the table, which isn't
+  // possible without key context.
+  if (arg.alloc_size() == this->alloc_size()) {
+    // No effective way for self-assignment to fall out of an efficient
+    // implementation so detect and bypass here.
+    if (&arg == this) {
+      return *this;
+    }
+    CARBON_DCHECK(arg.storage() != this->storage());
+    // The sizes match, so just clear out old entries.
+    if constexpr (!EntryT::IsTriviallyDestructible) {
+      this->view_impl_.ForEachEntry([](EntryT& entry) { entry.Destroy(); },
+                                    [](auto...) {});
     }
```

Comment:
Do we care about the case where these destructor calls modify either `*this` or `arg`? (I think and hope the answer is no.)

Reply:
Nope. At least, my intent is not to permit that. =]